### PR TITLE
Ebranchformer

### PR DESCRIPTION
**Pull request by KarelVesely84**
Hello Fangyun @csukuangfj ,
i did extend sherpa-onnx to support our EBranchformer encoder implementation 
that we currently use widely at Brno University in Technology.

The EBranchformer code is based on Conformer `model` from `transformers`, but the internals are different:
https://github.com/BUTSpeechFIT/huggingface_asr/blob/streaming_karel/src/models/encoders/e_branchformer.py

This allows to pre-train the encoder with BestRQ alg. and then fine-tune with modified icefall.
We would like to deploy it as a production system for streaming ASR (it already works for me locally).

So for us it would be good to have the support directly inside sherpa-onnx,
so we can use the official sherpa-onnx builds.

On the other hand, it is a bit specific model, not yet widespread.
Would you agree on accepting this extension into the codebase ?
Or, should we rely on our cusom builds ?

The Encoder assumes little different preset of input features derived from 
[Speech2TextFeatureExtractor](https://github.com/huggingface/transformers/blob/9e3a1072c25732d351a09218d8e8843148ecc3c9/src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py#L36), hence newly surfacing the FBANK options:
`normalize_samples` and `snip_edges`.

Best regards,
Karel

### CODE REVIEW

**Comment by csukuangfj** [comment]
Thanks! Will review it today.

File: sherpa-onnx/csrc/features.cc [hunk 1]
```diff
@@ -48,7 +48,9 @@ std::string FeatureExtractorConfig::ToString() const {
   os << "feature_dim=" << feature_dim << ", ";
   os << "low_freq=" << low_freq << ", ";
   os << "high_freq=" << high_freq << ", ";
-  os << "dither=" << dither << ")";
+  os << "dither=" << dither << ", ";
+  os << "normalize_samples=" << normalize_samples << ", ";
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/features.cc]:
```suggestion
  os << "normalize_samples=" << (normalize_samples ? "True" : "False" )<< ", ";
```

File: sherpa-onnx/csrc/features.cc [hunk 2]
```diff
@@ -48,7 +48,9 @@ std::string FeatureExtractorConfig::ToString() const {
...
+  os << "snip_edges=" << snip_edges << ")";
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/features.cc]:
```suggestion
  os << "snip_edges=" << (snip_edges ? "True" : "False") << ")";
```

File: sherpa-onnx/python/csrc/online-stream.cc [hunk 3]
```diff
@@ -34,6 +51,9 @@ void PybindOnlineStream(py::module *m) {
           py::arg("sample_rate"), py::arg("waveform"), kAcceptWaveformUsage,
           py::call_guard<py::gil_scoped_release>())
       .def("input_finished", &PyClass::InputFinished,
+           py::call_guard<py::gil_scoped_release>())
+      .def("get_frames", &PyClass::GetFrames,
+           py::arg("frame_index"), py::arg("n"),
```
Comment (csukuangfj) [inline sherpa-onnx/python/csrc/online-stream.cc]:
```suggestion
           py::arg("frame_index"), py::arg("n"), kGetFramesUsage
```

so that if you use `help(OnlineStream.get_frames)`, you can view the help info in python.

File: sherpa-onnx/csrc/online-transducer-model.cc [hunk 4]
```diff
@@ -92,6 +96,8 @@ std::unique_ptr<OnlineTransducerModel> OnlineTransducerModel::Create(
     const auto &model_type = config.model_type;
     if (model_type == "conformer") {
       return std::make_unique<OnlineConformerTransducerModel>(config);
+    } else if (model_type == "ebranchformer") {
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-transducer-model.cc:99]: Can you also update
https://github.com/k2-fsa/sherpa-onnx/blob/41d378a128312acc6da09251dc32891a4922f6ee/sherpa-onnx/csrc/online-transducer-model.cc#L180-L181

File: sherpa-onnx/csrc/online-transducer-model.cc [hunk 5]
```diff
@@ -115,6 +121,8 @@ std::unique_ptr<OnlineTransducerModel> OnlineTransducerModel::Create(
   switch (model_type) {
     case ModelType::kConformer:
       return std::make_unique<OnlineConformerTransducerModel>(config);
+    case ModelType::kEbranchformer:
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-transducer-model.cc:124]: Can you also update
https://github.com/k2-fsa/sherpa-onnx/blob/41d378a128312acc6da09251dc32891a4922f6ee/sherpa-onnx/csrc/online-transducer-model.cc#L199-L200

It is for Android and HarmonyOS.

**Review by csukuangfj — COMMENTED** [review]
Thanks!

Looks great to me. Left only some minor comments

**Comment by csukuangfj** [comment]
(Please ignore the failed CI tests.)

**Comment by KarelVesely84** [comment]
ok, good, thank you for the feedback
the remarks are integrated into the PR code now

**Review by csukuangfj — APPROVED** [review]
Thank you for your contribution!
